waspc/src/Wasp/Generator/NpmInstall.hs

```diff
@@ -0,0 +1,106 @@
+module Wasp.Generator.NpmInstall
+  ( ensureNpmInstall,
+  )
+where
+
+import Control.Concurrent (Chan, newChan, readChan)
+import Control.Concurrent.Async (concurrently)
+import Control.Monad (when)
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as B
+import StrongPath (Abs, Dir, File', Path', Rel, relfile, (</>))
+import qualified StrongPath as SP
+import System.Directory (doesFileExist, removeFile)
+import System.Exit (ExitCode (..))
+import Wasp.AppSpec (AppSpec)
+import Wasp.Generator.Common (ProjectRootDir)
+import qualified Wasp.Generator.Job as J
+import Wasp.Generator.Job.IO (printPrefixedJobMessage)
+import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
+import qualified Wasp.Generator.NpmDependencies as N
+import Wasp.Generator.ServerGenerator as SG
+import qualified Wasp.Generator.ServerGenerator.Setup as ServerSetup
+import Wasp.Generator.WebAppGenerator as WG
+import qualified Wasp.Generator.WebAppGenerator.Setup as WebAppSetup
+
+-- | Run npm install if needed
+ensureNpmInstall :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+ensureNpmInstall spec dstDir = do
+  let appSpecFullStackPackageJsonDependencies = N.buildFullStackNpmDependencies spec SG.waspNpmDependencies WG.waspNpmDependencies
+  case appSpecFullStackPackageJsonDependencies of
+    Left message -> return ([], [GenericGeneratorError ("npm install failed: " ++ message)])
+    Right fullStackDependencies -> do
+      needed <- isNpmInstallNeeded fullStackDependencies dstDir
+      if needed
+        then executeNpmInstall fullStackDependencies dstDir
+        else return ([], [])
+
+-- Run npm install for desired AppSpec dependencies
+executeNpmInstall :: N.FullStackNpmDependencies -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+executeNpmInstall appSpecFullStackPackageJsonDependencies dstDir = do
+  npmInstallResult <- installNpmDependencies dstDir
+  -- in case anything fails during installation that would leave node modules in
+  -- a broken state, we remove the file before we start npm install
+  fileExists <- doesFileExist dependenciesInstalledFp
+  when fileExists $ removeFile dependenciesInstalledFp
+  case npmInstallResult of
+    Left npmInstallError -> do
+      return ([], [GenericGeneratorError $ "npm install failed: " ++ npmInstallError])
+    Right () -> do
+      -- on successful npm install, record what we wanted to install
+      B.writeFile dependenciesInstalledFp (Aeson.encode appSpecFullStackPackageJsonDependencies)
+      return ([], [])
+  where
+    dependenciesInstalledFp = SP.fromAbsFile $ dstDir </> installedFullStackNpmDependenciesFileInProjectRootDir
+
+isNpmInstallNeeded :: N.FullStackNpmDependencies -> Path' Abs (Dir ProjectRootDir) -> IO Bool
```

Comment:
This also makes assumption stated above, that appSpec deps match the state on the disk.